src/App.tsx

```diff
@@ -15,36 +27,118 @@ export const goodsFromServer = [
   'Garlic',
 ];
 
+const goodsWithKeys = goodsFromServer.map((good, i) => {
+  return { id: i + 1, product: good };
+});
+
+type Action =
+  | { type: SortType.Alphabetical }
+  | { type: SortType.Length }
+  | { type: SortType.Reverse }
+  | { type: SortType.Clear };
+
+type State = {
+  sortedList: ProductWithKey[];
+  sortedBy: string;
+  isReversed: boolean;
+};
+
 export const App: React.FC = () => {
+  const initialState = {
+    sortedList: [...goodsWithKeys],
+    sortedBy: '',
+    isReversed: false,
+    isChanged: false,
+  };
+
+  function reducer(state: State, action: Action): State {
+    let { sortedList, sortedBy, isReversed } = state;
+
+    switch (action.type) {
+      case SortType.Alphabetical:
+        sortedBy = SortType.Alphabetical;
+        sortedList = !isReversed
+          ? [...sortedList].sort((a, b) => a.product.localeCompare(b.product))
+          : [...sortedList].sort((a, b) => b.product.localeCompare(a.product));
+        break;
+      case SortType.Length:
+        sortedBy = SortType.Length;
+        sortedList = !isReversed
+          ? [...sortedList].sort((a, b) => a.product.length - b.product.length)
+          : [...sortedList].sort((a, b) => b.product.length - a.product.length);
+        break;
+      case SortType.Reverse:
+        isReversed = !isReversed;
+        sortedList = sortedList.reverse();
+        break;
+      case SortType.Clear:
+        sortedBy = '';
+        isReversed = false;
+        sortedList = [...goodsWithKeys];
+        break;
+      default:
+        break;
+    }
+
+    return { sortedList, sortedBy, isReversed };
+  }
+
+  const [{ sortedList, sortedBy, isReversed }, dispatch] = useReducer(
+    reducer,
+    initialState,
+  );
+
   return (
     <div className="section content">
       <div className="buttons">
-        <button type="button" className="button is-info is-light">
+        <button
+          onClick={() => dispatch({ type: SortType.Alphabetical })}
+          type="button"
+          className={
+            sortedBy !== SortType.Alphabetical
+              ? 'button is-info is-light'
+              : 'button is-info'
```

Comment:
[UX] - To improve user experience, consider marking the active button more clearly to indicate the current sorting action. For instance, you could use a different style or icon to highlight the active state.